### PR TITLE
Transformer (Petr)

### PR DESCRIPTION
**Pull request by laitoch**
Another take on adding in Transformer.

---

TODOs:
- `SimpleTransformerEncoder` masking
- Write `SimpleTransformerDecoder` (blocked by "Understanting TODOs")
---

Understanding TODOs (They are blockers for this PR, but should be added as comments into [here](https://github.com/pixelneo/dialogue-transformer-e2e/pull/10)):
- Document what exactly are the inputs and outputs of the `BSpanDecoder` and `ResponseDecoder` `forward` functions.
- Understand `get_sparse_input_aug` and `get_sparse_selective_input`. Do we need to use/re-implement them?

### CODE REVIEW

File: implementation/transformer/transformer.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+        - Tutorials:
+            - https://pytorch.org/tutorials/beginner/transformer_tutorial.html
+            - https://andrewpeng.dev/transformer-pytorch/
+            - https://nlp.seas.harvard.edu/2018/04/03/attention.html
+    """
+    def __init__(self, vocabulary_size, dropout):
+        super().__init__()
+
+        # TODO: Do we want to use the dropout functionality of
+        # PositionalEncoder, TransformerEncoderLayer, or both?
```
Comment (pixelneo) [inline implementation/transformer/transformer.py:42]: I'd leave it for now. In the original paper, they use both but i do not think it has a high priority here.

File: implementation/transformer/transformer.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+            max_len=self.cfg.max_ts
+        )
+
+        self.transformer_encoder = nn.modules.TransformerEncoder(
+            nn.modules.TransformerEncoderLayer(
+                d_model=D_MODEL,  # 512 used in the original paper.
+                nhead=8,  # Multi-head attention head count. 8 used in the original paper.
+                dropout=dropout,
+            ),
+            num_layers=6  # Number of sub-encoder-layers in the encoder. 6 used in the original paper.
```
Comment (pixelneo) [inline implementation/transformer/transformer.py:76]: We will need much smaller dimensions, otherwise it both will overfit heavily (in original paper, they train it on MT, 4.5M pairs) and it will be computationally demanding (orig. paper - 8 P100 GPUs, 0.4s per training step).

**Comment by pixelneo** [comment]
> Understand get_sparse_input_aug and get_sparse_selective_input. Do we need to use/re-implement them?

I've added an explaantion of this as a comment [here](https://github.com/pixelneo/dialogue-transformer-e2e/pull/10)
